network/hnswrapper/hnsv2wrapperwithtimeout.go

```diff
@@ -0,0 +1,344 @@
+package hnswrapper
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Microsoft/hcsshim/hcn"
+)
+
+type Hnsv2wrapperwithtimeout struct {
+	Hnsv2 HnsV2WrapperInterface
+}
+
+type CreateEndpointFuncResult struct {
+	endpoint *hcn.HostComputeEndpoint
+	Err      error
+}
+
+type GetEndpointByIDFuncResult struct {
+	endpoint *hcn.HostComputeEndpoint
+	Err      error
+}
+
+type ListEndpointsFuncResult struct {
+	endpoints []hcn.HostComputeEndpoint
+	Err       error
+}
+
+type CreateNetworkFuncResult struct {
+	network *hcn.HostComputeNetwork
+	Err     error
+}
+
+type GetNamespaceByIDFuncResult struct {
+	namespace *hcn.HostComputeNamespace
+	Err       error
+}
+
+type GetNetworkByNameFuncResult struct {
+	network *hcn.HostComputeNetwork
+	Err     error
+}
+
+type GetNetworkByIDFuncResult struct {
+	network *hcn.HostComputeNetwork
+	Err     error
+}
+
+func (h Hnsv2wrapperwithtimeout) CreateEndpoint(endpoint *hcn.HostComputeEndpoint) (*hcn.HostComputeEndpoint, error) {
+
+	r := make(chan CreateEndpointFuncResult)
+
+	go func() {
+		endpoint, err := h.Hnsv2.CreateEndpoint(endpoint)
+
+		r <- CreateEndpointFuncResult{
+			endpoint: endpoint,
+			Err:      err,
+		}
+	}()
+
+	// Listen on our channel AND a timeout channel - which ever happens first.
+	select {
+	case res := <-r:
+		return res.endpoint, res.Err
+	case <-time.After(3 * time.Second):
```

Comment:
alternatively use context.Context and cancel that context withTimeout

https://www.sohamkamani.com/golang/context-cancellation-and-values/#emitting-a-cancellation-event

Reply:
updated